src/webots/control/WbControlledWorld.cpp

```diff
@@ -219,9 +219,22 @@ void WbControlledWorld::addControllerConnection() {
       robotName = buffer;
       delete[] buffer;
     }
+    int nbExternalController = 0;
     foreach (WbRobot *const robot, robots()) {
-      if (robot->isControllerStarted())
-        continue;
+      if (robot->controllerName() == "<extern>")
+	    nbExternalController++;
+    }
+    foreach (WbRobot *const robot, robots()) {
+      if (robot->isControllerStarted()) {
+        // if a specific robot have been targeted by giving a name, or only one robot is in the scene (hence it's the target)
+        if (robot->controllerName() == "<extern>" && (robotName == robot->name() || nbExternalController == 1)) {
+          // automatically disconnect from previous extern controller that may have failed, the slot could be immediatly used to restart
+          WbLog::info(tr("Closing extern controller connection for robot \"%1\".").arg(robot->name()));
+          updateRobotController(robot);
+        }
+        else // this controller seems to be in active use, ignore it
+          continue;
+      }
```

Comment:
The two for-loops can be merged into a single one so that if it is extern and the robot name matches the loop can exit immediately.

Reply:
To be fair, I'm confused : 
As I understand it, you advice to move lignes 224-225
```
      if (robot->controllerName() == "<extern>")
	    nbExternalController++;
```
inside and at the start of the foreach starting at 227
I'm right ? 
However, let's consider the following case : 
2 robots (A and B in this order) with external controllers (resp c_A, c_B), both started, and robotName == "B" (user want to reconnect c_B)
On the first loop, we would get : on line 230 : `nbExternalController == 1` (because the `if (robot->controllerName() == "<extern>")` check have been placed at the beginning of the loop)
This will allows the `if` to trigger the reconnect of c_B on robot A (and kicking c_A), which is not what we want.
This is why we need to know in advance how many external we have, to be able to trigger this special case of only one external controller (ie`nbExternalController == 1` ) correctly.
I do think that two separate loops are the way to go. Did I get it right ?

Reply:
What I mean is, in the for loop, count the extern controllers and save the WbRobot instance that matches so that disconnect and restart can be applied just after the loop:
* if `WEBOTS_ROBOT_NAME` is set, you don't need to count the extern controller, you can store the current WbRobot instance and break immediately the loop
* if `WEBOTS_ROBOT_NAME` is empty you store the WbRobot instance of the first started extern controller but continue within the loop
After the loop is over you can use the stored WbRobot instance to apply the disconnect/controller start.

Reply:
Well, that not went that well... I've been trying to implement what you was explaining, but even if I succeeded, I ended up with wacky code and strange construct.
If I find some time tomorrow I will try to think a bit more about it.

Reply:
I must say that I've not "clean" idea...
I would be interesting to see a nicer solution, but for the moment I prefer just request review again, to not stale the PR.